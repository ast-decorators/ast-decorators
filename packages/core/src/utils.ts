import {ASTDecoratorTransformerOptions} from '@ast-decorators/typings';
import {BabelFileResult, NodePath} from '@babel/core';
import {Binding} from '@babel/traverse';
import {
  CallExpression,
  Decorator,
  Identifier,
  ImportDeclaration,
  ImportDefaultSpecifier,
  ImportNamespaceSpecifier,
  ImportSpecifier,
  isCallExpression,
  isMemberExpression,
  MemberExpression,
  StringLiteral,
} from '@babel/types';

export type ASTDecoratorExclusionOptions = {
  paths?: string[];
  names?: RegExp | string[];
};

export type ASTDecoratorCoreOptions = Readonly<{
  exclude?: ASTDecoratorExclusionOptions;
  transformers?: ASTDecoratorTransformerOptions;
}>;

export type PluginPass<T> = Readonly<{
  cwd: string;
  file: BabelFileResult;
  filename?: string;
  key: string;
  opts?: T;
}>;

const $args = Symbol('args');
const $binding = Symbol('binding');
const $decorator = Symbol('decorators');
const $id = Symbol('id');
const $isCall = Symbol('isCall');
const $object = Symbol('object');

const $handleDecoratorWithArgs = Symbol('handleDecoratorWithArgs');
const $handleDecoratorWithoutArgs = Symbol('handleDecoratorWithoutArgs');
const $handleMemberDecorator = Symbol('handleMemberDecorator');

export default class DecoratorMetadata {
  private [$args]?: readonly NodePath[];
  private [$binding]?: Binding;
  private [$decorator]: NodePath<Decorator>;
  private [$id]: NodePath<Identifier>;
  private [$isCall]: boolean;
  private [$object]?: NodePath<Identifier>;

  public constructor(decorator: NodePath<Decorator>) {
    this[$decorator] = decorator;
    const expression = decorator.get('expression');

    this[$isCall] = isCallExpression(expression);

    if (this[$isCall]) {
      this[$handleDecoratorWithArgs](expression as NodePath<CallExpression>);
    } else {
      this[$handleDecoratorWithoutArgs](
        expression as NodePath<MemberExpression | Identifier>,
      );
    }

    if (this.isFree) {
      this[$binding] = decorator.scope.getBinding(this[$id]!.node.name);
    } else {
      this[$binding] = decorator.scope.getBinding(this[$object]!.node.name);
    }
  }

  public get args(): readonly NodePath[] {
    return this[$args] ?? [];
  }

  public get binding(): Binding | undefined {
    return this[$binding];
  }

  public get bindingId(): NodePath<Identifier> {
    return this.isFree ? this[$id] : this[$object]!;
  }

  public get importSpecifier():
    | NodePath<
        ImportSpecifier | ImportNamespaceSpecifier | ImportDefaultSpecifier
      >
    | undefined {
    return this[$binding] ? (this[$binding]!.path as any) : undefined;
  }

  public get importSource(): NodePath<StringLiteral> | undefined {
    const {importSpecifier} = this;

    if (!importSpecifier) {
      return undefined;
    }

    const declaration = importSpecifier.parentPath as NodePath<
      ImportDeclaration
    >;

    return declaration.get('source');
  }

  public get isCall(): boolean {
    return this[$isCall];
  }

  /**
   * Is decorator identifier a part of MemberExpression?
   */
  public get isFree(): boolean {
    return !this[$object];
  }

  public get property(): string | undefined {
    return this.isFree ? undefined : this[$id].node.name;
  }

  public removeDecorator(): void {
    this[$decorator].remove();
  }

  public removeBinding(): void {
    if (!this[$binding]) {
      return;
    }

    this[$binding]!.referencePaths = this[$binding]!.referencePaths.filter(
      p => p !== this.bindingId,
    );

    if (this[$binding]!.referencePaths.length === 0) {
      const declaration = this.importSpecifier!.parentPath as NodePath<
        ImportDeclaration
      >;

      this.importSpecifier!.remove();

      if (declaration.node.specifiers.length === 0) {
        declaration.remove();
      }
    }
  }

  private [$handleDecoratorWithArgs](
    expression: NodePath<CallExpression>,
  ): void {
    this[$args] = expression.get('arguments');

    const callee = expression.get('callee');

    if (isMemberExpression(callee)) {
      this[$handleMemberDecorator](callee as NodePath<MemberExpression>);
    } else {
      this[$id] = callee as NodePath<Identifier>;
    }
  }

  private [$handleDecoratorWithoutArgs](
    expression: NodePath<MemberExpression | Identifier>,
  ): void {
    if (isMemberExpression(expression)) {
      this[$handleMemberDecorator](expression as NodePath<MemberExpression>);
    } else {
      this[$id] = expression as NodePath<Identifier>;
    }
  }

  private [$handleMemberDecorator](
    expression: NodePath<MemberExpression>,
  ): void {
    this[$object] = expression.get('object') as NodePath<Identifier>;
    this[$id] = expression.get('property') as NodePath<Identifier>;
  }
}
