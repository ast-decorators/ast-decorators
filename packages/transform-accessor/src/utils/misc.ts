import ASTDecoratorsError from '@ast-decorators/utils/lib/ASTDecoratorsError';
import {SuitabilityFactors} from '@ast-decorators/utils/lib/checkSuitability';
import {
  ASTClassMemberDecorator,
  ClassMemberMethod,
  ClassMemberProperty,
  PluginPass,
  PrivacyType,
} from '@ast-decorators/utils/lib/common';
import createPropertyByPrivacy from '@ast-decorators/utils/lib/createPropertyByPrivacy';
import getMemberName from '@ast-decorators/utils/lib/getMemberName';
import {NodePath, template} from '@babel/core';
import {
  ArrowFunctionExpression,
  callExpression,
  CallExpression,
  Class,
  cloneNode,
  Decorator,
  functionDeclaration,
  FunctionExpression,
  identifier,
  Identifier,
  isArrowFunctionExpression,
  isClassPrivateProperty,
  isClassProperty,
  isFunctionExpression,
  isIdentifier,
  isMemberExpression,
  memberExpression,
  MemberExpression,
  PrivateName,
  ThisExpression,
  thisExpression,
} from '@babel/types';
import shouldUseContext from './context';

export type TransformAccessorOptions = Readonly<{
  interceptorContext?: InterceptorContext;
  privacy?: PrivacyType;
  singleAccessorDecorators?: SuitabilityFactors;
  transformerPath?: string;
  useClassNameForStatic?: boolean;
}>;

export type AccessorInterceptor = (value: any) => any;
export type AccessorInterceptorNode =
  | FunctionExpression
  | ArrowFunctionExpression
  | Identifier
  | MemberExpression;

export type InterceptorContext = Readonly<{
  disableByDefault?: boolean;
  exclude?: SuitabilityFactors;
}>;

export type AccessorMethodCreatorOptions = Readonly<{
  useClassName: boolean;
  useContext: boolean;
  preservingDecorators: Decorator[] | null;
}>;

export type AccessorMethodCreator = (
  klass: NodePath<Class>,
  member: NodePath<ClassMemberProperty>,
  accessor: NodePath<AccessorInterceptorNode> | undefined,
  storage: Identifier | PrivateName,
  options: AccessorMethodCreatorOptions,
) => ClassMemberMethod;

export const assert = (
  decorator: string,
  member: NodePath<ClassMemberProperty>,
  interceptors: ReadonlyArray<NodePath<AccessorInterceptorNode> | undefined>,
): void => {
  if (!isClassProperty(member) && !isClassPrivateProperty(member)) {
    throw new ASTDecoratorsError(
      `Applying @${decorator} decorator to something other than property is not allowed`,
    );
  }

  for (const interceptor of interceptors) {
    if (
      interceptor &&
      !isFunctionExpression(interceptor) &&
      !isArrowFunctionExpression(interceptor) &&
      !isIdentifier(interceptor) &&
      !isMemberExpression(interceptor)
    ) {
      throw new ASTDecoratorsError(
        'Accessor interceptor can only be function, free variable or object property',
      );
    }
  }
};

export const createStorage = (
  klass: NodePath<Class>,
  member: NodePath<ClassMemberProperty>,
  privacy: PrivacyType = 'hard',
): ClassMemberProperty =>
  createPropertyByPrivacy(privacy, String(getMemberName(member.node)), klass, {
    // @ts-ignore
    static: member.node.static,
    value: member.node.value,
  });

const declarator = template(`const VAR = FUNCTION`);

export const generateInterceptor = (
  klass: NodePath<Class>,
  interceptor: AccessorInterceptorNode,
  type: 'get' | 'set',
): Identifier => {
  const isArrow = isArrowFunctionExpression(interceptor);
  const isRegular = isFunctionExpression(interceptor);

  const accessorId =
    isArrow || isRegular
      ? klass.parentPath.scope.generateUidIdentifier(`${type}Interceptor`)
      : (interceptor as Identifier);

  if (isArrow) {
    klass.insertBefore(
      declarator({
        FUNCTION: interceptor,
        VAR: accessorId,
      }),
    );
  } else if (isRegular) {
    klass.insertBefore(
      functionDeclaration(
        accessorId,
        (interceptor as FunctionExpression).params,
        (interceptor as FunctionExpression).body,
      ),
    );
  }

  return accessorId;
};

export const createAccessorDecorator = (
  decorator: string,
  interceptor: NodePath<AccessorInterceptorNode> | undefined,
  impl: AccessorMethodCreator,
): ASTClassMemberDecorator<TransformAccessorOptions> => (
  klass: NodePath<Class>,
  member: NodePath<ClassMemberProperty>,
  {
    interceptorContext,
    privacy,
    useClassNameForStatic,
  }: TransformAccessorOptions = {},
  {filename}: PluginPass,
): void => {
  assert(decorator, member, [interceptor]);

  const storage = createStorage(klass, member, privacy);

  const method = impl(
    klass,
    member,
    interceptor,
    storage.key as Identifier | PrivateName,
    {
      preservingDecorators: member.node.decorators,
      // @ts-ignore
      useClassName: !!member.node.static && !!useClassNameForStatic,
      useContext: shouldUseContext(interceptor, interceptorContext, filename),
    },
  );

  member.replaceWithMultiple([storage, method]);
};

export const ownerNode = (
  klass: NodePath<Class>,
  useClassName: boolean,
): Identifier | ThisExpression =>
  useClassName && klass.node.id ? cloneNode(klass.node.id) : thisExpression();

export const injectInterceptor = (
  klass: NodePath<Class>,
  interceptor: AccessorInterceptorNode,
  value: MemberExpression | Identifier,
  type: 'get' | 'set',
  useContext: boolean,
  useClassName: boolean,
): CallExpression => {
  const interceptorId = generateInterceptor(klass, interceptor, type);

  return isArrowFunctionExpression(interceptor) ||
    (isIdentifier(interceptor) && !useContext)
    ? callExpression(interceptorId, [value])
    : callExpression(memberExpression(interceptorId, identifier('call')), [
        ownerNode(klass, useClassName),
        value,
      ]);
};
