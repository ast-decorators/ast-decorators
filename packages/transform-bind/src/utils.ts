import {ClassMember, ClassMemberMethod} from '@ast-decorators/typings';
import ASTDecoratorsError from '@ast-decorators/utils/lib/ASTDecoratorsError';
import {NodePath} from '@babel/traverse';
import {
  callExpression,
  Class,
  ClassMethod,
  ClassPrivateMethod,
  classPrivateProperty,
  classProperty,
  FunctionDeclaration,
  functionDeclaration,
  identifier,
  isClassMethod,
  isClassPrivateMethod,
  memberExpression,
  thisExpression,
} from '@babel/types';

export type TransformBindOptions = Readonly<{
  transformerPath?: string;
}>;

export const assert = (decorator: string, member: ClassMember): void => {
  if (!isClassMethod(member) && !isClassPrivateMethod(member)) {
    throw new ASTDecoratorsError(
      `Applying @${decorator} decorator to something other than method is not allowed`,
    );
  }
};

const applyBindingToClassPrivateMethods = (
  klass: NodePath<Class>,
  methodsToBind: ReadonlyArray<NodePath<ClassPrivateMethod>>,
): void => {
  const methodFunctions: FunctionDeclaration[] = [];
  for (const method of methodsToBind) {
    const {async, body, generator, key, params} = method.node;

    const functionId = klass.parentPath.scope.generateUidIdentifier(
      key.id.name,
    );

    methodFunctions.push(
      functionDeclaration(functionId, params, body, generator, async),
    );

    method.replaceWith(
      classPrivateProperty(
        key,
        callExpression(memberExpression(functionId, identifier('bind')), [
          thisExpression(),
        ]),
      ),
    );
  }

  klass.insertBefore(methodFunctions);
};

const applyBindingToClassMethods = (
  _: NodePath<Class>,
  methodsToBind: ReadonlyArray<NodePath<ClassMethod>>,
): void => {
  for (const method of methodsToBind) {
    const {computed, key, static: _static} = method.node;

    const bindingExpression = callExpression(
      memberExpression(
        memberExpression(thisExpression(), key),
        identifier('bind'),
      ),
      [thisExpression()],
    );

    method.insertBefore(
      classProperty(key, bindingExpression, null, null, computed, _static),
    );
  }
};

export const applyBinding = (
  klass: NodePath<Class>,
  methodsToBind: ReadonlyArray<NodePath<ClassMemberMethod>>,
): void => {
  const methods: Array<NodePath<ClassMethod>> = [];
  const privateMethods: Array<NodePath<ClassPrivateMethod>> = [];

  for (const method of methodsToBind) {
    if (isClassPrivateMethod(method)) {
      privateMethods.push(method as NodePath<ClassPrivateMethod>);
    } else {
      methods.push(method as NodePath<ClassMethod>);
    }
  }

  applyBindingToClassMethods(klass, methods);
  applyBindingToClassPrivateMethods(klass, privateMethods);
};
