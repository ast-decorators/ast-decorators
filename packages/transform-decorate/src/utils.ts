export type TransformBindOptions = Readonly<{
  transformerPath?: string;
}>;
