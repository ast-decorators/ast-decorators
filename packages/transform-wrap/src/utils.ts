import ASTDecoratorsError from '@ast-decorators/utils/lib/ASTDecoratorsError';
import getMemberName from '@ast-decorators/utils/lib/getMemberName';
import type {ClassMemberMethod} from '@ast-decorators/utils/src/common';
import {
  ArrowFunctionExpression,
  FunctionExpression,
  Identifier,
  isClassPrivateProperty,
  isClassProperty,
  isLiteral,
  MemberExpression,
} from '@babel/types';

export type TransformWrapOptions = Readonly<{
  transformerPath?: string;
}>;

export type AllowedWrappers =
  | FunctionExpression
  | ArrowFunctionExpression
  | Identifier
  | MemberExpression;

export const assertWrap = (
  wrapperFunction: AllowedWrappers | undefined,
  member: ClassMemberMethod,
): void => {
  if (!wrapperFunction) {
    const methodName = getMemberName(member);

    throw new ASTDecoratorsError(
      `@wrap: No wrapper function provided${
        methodName !== undefined ? ` for ${methodName}` : ''
      }`,
    );
  }

  if (
    (isClassProperty(member) || isClassPrivateProperty(member)) &&
    (!member.value || isLiteral(member.value))
  ) {
    throw new ASTDecoratorsError(
      `@wrap can only be applied to class methods or properties with ` +
        'functions assigned',
    );
  }
};
