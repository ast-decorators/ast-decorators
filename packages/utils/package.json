{
  "name": "@ast-decorators/utils",
  "version": "0.1.0",
  "description": "Common utils for AST decorators",
  "types": "lib/index.d.ts",
  "scripts": {
    "build": "npm run build:remove && run-p build:js build:dts",
    "build:js": "babel src -d lib -x .ts --ignore src/**/*.d.ts --copy-files --config-file ../../babel.config.json",
    "build:dts": "tsc -p tsconfig.build.json --emitDeclarationOnly",
    "build:remove": "rimraf lib",
    "typecheck": "tsc -p tsconfig.json --noEmit"
  },
  "dependencies": {
    "@ast-decorators/typings": "^0.1.0",
    "@babel/types": "^7"
  }
}
