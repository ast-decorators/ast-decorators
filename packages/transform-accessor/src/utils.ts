import ASTDecoratorsError from '@ast-decorators/utils/lib/ASTDecoratorsError';
import type {SuitabilityFactors} from '@ast-decorators/utils/lib/checkSuitability';
import type {
  ClassMemberMethod,
  ClassMemberProperty,
  PrivacyType,
} from '@ast-decorators/utils/lib/common';
import createPropertyByPrivacy from '@ast-decorators/utils/lib/createPropertyByPrivacy';
import getMemberName from '@ast-decorators/utils/lib/getMemberName';
import {ClassMember} from '@ast-decorators/utils/src/common';
import template from '@babel/template';
import type {NodePath, Scope} from '@babel/traverse';
import {
  ArrayPattern,
  ArrowFunctionExpression,
  Class,
  cloneNode,
  Decorator,
  FunctionDeclaration,
  functionDeclaration,
  FunctionExpression,
  Identifier,
  isArrayPattern,
  isArrowFunctionExpression,
  isClassMethod,
  isClassPrivateMethod,
  isClassPrivateProperty,
  isClassProperty,
  isFunctionExpression,
  isIdentifier,
  isMemberExpression,
  isObjectPattern,
  MemberExpression,
  ObjectPattern,
  PrivateName,
  ThisExpression,
  thisExpression,
  variableDeclaration,
  VariableDeclaration,
  variableDeclarator,
} from '@babel/types';

export type TransformedNode = readonly [
  ClassMemberMethod,
  ReadonlyArray<FunctionDeclaration | VariableDeclaration>,
];

export type TransformAccessorOptions = Readonly<{
  privacy?: PrivacyType;
  singleAccessorDecorators?: SuitabilityFactors;
  transformerPath?: string;
  useClassNameForStatic?: boolean;
}>;

export type AccessorInterceptor = (value: any) => any;
export type AccessorInterceptorNode =
  | FunctionExpression
  | ArrowFunctionExpression
  | Identifier
  | MemberExpression;

export type AccessorMethodCreatorOptions = Readonly<{
  useClassName: boolean;
  preservingDecorators: Decorator[] | null;
}>;

export type AccessorMethodCreator = (
  klass: NodePath<Class>,
  member: NodePath<ClassMember>,
  interceptor: NodePath<AccessorInterceptorNode> | undefined,
  storage: Identifier | PrivateName | undefined,
  options: AccessorMethodCreatorOptions,
) => TransformedNode | null;

export const isGetter = (member: ClassMemberMethod): boolean =>
  member.kind === 'get';
export const isSetter = (member: ClassMemberMethod): boolean =>
  member.kind === 'set';

export const assert = (
  decorator: string,
  member: ClassMember,
  interceptors: ReadonlyArray<AccessorInterceptorNode | undefined>,
): void => {
  if (
    !(isClassProperty(member) || isClassPrivateProperty(member)) &&
    !(
      (isClassMethod(member) || isClassPrivateMethod(member)) &&
      (isGetter(member) || isSetter(member))
    )
  ) {
    throw new ASTDecoratorsError(
      `Applying @${decorator} decorator to something other than property or accessor is not allowed`,
    );
  }

  for (const interceptor of interceptors) {
    if (
      interceptor &&
      !isFunctionExpression(interceptor) &&
      !isArrowFunctionExpression(interceptor) &&
      !isIdentifier(interceptor) &&
      !isMemberExpression(interceptor)
    ) {
      throw new ASTDecoratorsError(
        'Accessor interceptor can only be function, free variable or object property',
      );
    }
  }
};

export const createStorage = (
  klass: NodePath<Class>,
  member: ClassMemberProperty,
  privacy: PrivacyType = 'hard',
): ClassMemberProperty =>
  createPropertyByPrivacy(privacy, getMemberName(member), klass, {
    // @ts-expect-error: "static" is not listed in d.ts
    static: member.static,
    value: member.value,
  });

const declarator = template(`const VAR = FUNCTION`);

export const ownerNode = (
  klass: Class,
  useClassName: boolean,
): Identifier | ThisExpression =>
  useClassName && klass.id ? cloneNode(klass.id) : thisExpression();

export const prepareInterceptor = (
  klass: NodePath<Class>,
  interceptor: AccessorInterceptorNode,
  type: 'get' | 'set',
): readonly [
  Identifier | MemberExpression,
  FunctionDeclaration | VariableDeclaration | null,
] => {
  const isArrow = isArrowFunctionExpression(interceptor);
  const isRegular = isFunctionExpression(interceptor);

  const interceptorId =
    isArrow || isRegular
      ? klass.parentPath.scope.generateUidIdentifier(type)
      : (interceptor as Identifier | MemberExpression);

  return [
    interceptorId,
    isArrow
      ? (declarator({
          FUNCTION: interceptor,
          VAR: interceptorId as Identifier,
        }) as VariableDeclaration)
      : isRegular
      ? functionDeclaration(
          interceptorId as Identifier,
          (interceptor as FunctionExpression).params,
          (interceptor as FunctionExpression).body,
        )
      : null,
  ];
};

export const unifyValueParameter = (
  scope: Scope,
  rawValue: Identifier | ArrayPattern | ObjectPattern,
): [Identifier, VariableDeclaration?] => {
  if (isObjectPattern(rawValue) || isArrayPattern(rawValue)) {
    const value = scope.generateUidIdentifier('value');

    return [
      value,
      variableDeclaration('const', [variableDeclarator(rawValue, value)]),
    ];
  }

  return [rawValue];
};
