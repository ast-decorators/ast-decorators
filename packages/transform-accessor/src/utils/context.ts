import ASTDecoratorsError from '@ast-decorators/utils/lib/ASTDecoratorsError';
import checkSuitability from '@ast-decorators/utils/lib/checkSuitability';
import {ImportMetadata} from '@ast-decorators/utils/lib/metadata';
import {NodePath} from '@babel/core';
import {
  Identifier,
  isArrowFunctionExpression,
  isFunctionDeclaration,
  isFunctionExpression,
  isImportDeclaration,
  isMemberExpression,
  isVariableDeclarator,
  MemberExpression,
} from '@babel/types';
import {AccessorInterceptorNode, InterceptorContext} from './misc';

const shouldUseContext = (
  fn: NodePath<AccessorInterceptorNode> | undefined,
  {disableByDefault = false, exclude = {}}: InterceptorContext = {},
  filename: string,
): boolean => {
  if (!fn) {
    return false;
  }

  // If it is a declared regular function, context should be used.
  if (isFunctionExpression(fn)) {
    return true;
  }

  // Is it is a declared arrow function, context is unnecessary.
  if (isArrowFunctionExpression(fn)) {
    return false;
  }

  const isMember = isMemberExpression(fn);

  const id = isMember
    ? (fn.get('object') as NodePath<Identifier>)
    : (fn as NodePath<Identifier>);

  const binding = fn.scope.getBinding(id.node.name);

  if (!binding) {
    throw new ASTDecoratorsError(`${id.node.name} is not defined`);
  }

  let name: string | undefined;
  let source: string | undefined;

  const {path: declarationOrSpecifier} = binding;

  if (isImportDeclaration(declarationOrSpecifier.parentPath)) {
    // If the callback is imported, let's use global transformer rules.

    const {importSource, originalImportName} = new ImportMetadata(
      fn as NodePath<MemberExpression | Identifier>,
    );

    name = originalImportName;
    source = importSource?.node.value;
  } else if (!isMember) {
    // If element is declared in the same file and it is an arrow or a regular
    // function, let's decide if we need to add context.

    if (isVariableDeclarator(declarationOrSpecifier)) {
      const init = declarationOrSpecifier.get('init');

      if (isFunctionExpression(init)) {
        return true;
      } else if (isArrowFunctionExpression(init)) {
        return false;
      }
    } else if (isFunctionDeclaration(declarationOrSpecifier)) {
      return true;
    }

    throw new ASTDecoratorsError(`${id.node.name} is not a function`);
  } else {
    // If our function is a part of an object, let's use global transformer
    // settings.
    ({name} = (fn.get('property') as NodePath<Identifier>).node);
  }

  const suitable = checkSuitability({name, source}, exclude, filename);

  // If context is disabled by default, all functions that fit "exclude" option
  // should use context. Otherwise, exclusions should not use context.
  return disableByDefault ? suitable : !suitable;
};

export default shouldUseContext;
