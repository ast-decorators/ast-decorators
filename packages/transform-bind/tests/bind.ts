import {compare as _compare} from '../../../utils/testing';
import commonOptions from './fixtures/options';

const compare = async (
  fixture: string,
  options?: string | object,
): Promise<void> => _compare(__dirname, 'bind', fixture, options);

describe('@ast-decorators/transform-bind', () => {
  describe('@bind', () => {
    it('compiles for regular method', async () => {
      await compare('default', commonOptions);
    });

    it('compiles for private method', async () => {
      await compare('private-method', commonOptions);
    });
  });
});
