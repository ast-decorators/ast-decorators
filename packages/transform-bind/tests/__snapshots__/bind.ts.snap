// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@ast-decorators/transform-bind @bind compiles for private method 1`] = `
"function _bar() {
  console.log(this);
}

class Foo {
  #bar = _bar.bind(this);
}"
`;

exports[`@ast-decorators/transform-bind @bind compiles for regular method 1`] = `
"class Foo {
  bar = this.bar.bind(this);

  bar() {
    console.log(this);
  }

}"
`;
